In [1]:
# loading libraries 

import matplotlib as plt
import pandas as pd
import os
import pycountry
import string
from unidecode import unidecode

path = os.getcwd()

In [2]:
# loading the data

ucdp = pd.read_csv(path + "/data/ucdp_grouped.csv")
gdelt = pd.read_csv(path + "/data/final_gdelt_bycountry.txt")

In [3]:
ucdp.head(10)

month  year                    adm_1                country  country_id  \
0      1  1989  Agusan del Sur province            Philippines         840   
1      1  1989  Alta Verapaz department              Guatemala          90   
2      1  1989        Ancash department                   Peru         135   
3      1  1989     Antioquia department               Colombia         100   
4      1  1989      Apurímac department                   Peru         135   
5      1  1989      Arequipa department                   Peru         135   
6      1  1989              Assam state                  India         750   
7      1  1989      Ayacucho department                   Peru         135   
8      1  1989           Azerbaijan SSR  Russia (Soviet Union)         365   
9      1  1989         Baghlan province            Afghanistan         700   

     region  best  best_state  best_nonstate  best_onesided  deaths_civilians  \
0      Asia     5         0.0            0.0            5.0                 5   
1  Americas     4         4.0            0.0            0.0                 0   
2  Americas     4         4.0            0.0            0.0                 0   
3  Americas     0         0.0            0.0            0.0                 0   
4  Americas    17        17.0            0.0            0.0                 0   
5  Americas     2         2.0            0.0            0.0                 0   
6      Asia    35        35.0            0.0            0.0                 0   
7  Americas    25        22.0            0.0            3.0                 3   
8    Europe    42         0.0           42.0            0.0                 0   
9      Asia   670       670.0            0.0            0.0                 0   

   number_of_sources  count_conflict_new_id  freq_conflict_new_id  \
0                0.0                      1                   495   
1                0.0                      1                   233   
2                0.0                      1                   292   
3                0.0                      1                   605   
4                0.0                      1                   292   
5                1.0                      1                   292   
6                2.0                      1                   421   
7                0.0                      2                   292   
8                0.0                      1                  4809   
9                0.0                      1                   333   

   count_dyad_new_id  freq_dyad_new_id  MonthYear  
0                  1               962     198901  
1                  1               469     198901  
2                  1               631     198901  
3                  1              1072     198901  
4                  1               631     198901  
5                  1               631     198901  
6                  1               884     198901  
7                  2               631     198901  
8                  1              5419     198901  
9                  1               724     198901

In [4]:
gdelt.head(10)

ActionGeo_ADM1Code isocode  year  month  count_events_1  count_events_2  \
0                 AA     ABW  1920      1             2.0             0.0   
1                 AA     ABW  1981     10             2.0             0.0   
2                 AA     ABW  1981     12             0.0             0.0   
3                 AA     ABW  1983      1             0.0             0.0   
4                 AA     ABW  1983      2             6.0             0.0   
5                 AA     ABW  1983      3             2.0             6.0   
6                 AA     ABW  1983      4             2.0             0.0   
7                 AA     ABW  1983      5             2.0             0.0   
8                 AA     ABW  1983     12             0.0             0.0   
9                 AA     ABW  1984      1             0.0             0.0   

   count_events_3  count_events_4  count_events_5  count_events_6  ...  \
0             0.0             2.0             0.0             2.0  ...   
1             2.0             0.0             0.0             0.0  ...   
2             0.0            22.0             0.0             0.0  ...   
3             0.0             0.0             0.0             0.0  ...   
4             0.0             0.0             0.0             0.0  ...   
5             6.0             0.0             0.0             0.0  ...   
6             0.0             8.0             0.0             0.0  ...   
7             4.0             0.0             2.0             0.0  ...   
8             0.0             4.0             0.0             0.0  ...   
9             0.0             4.0             0.0             6.0  ...   

   count_events_20_opp  Actor1Code_count  Actor1CountryCode_count  \
0                  0.0               0.0                      0.0   
1                  0.0               0.0                      0.0   
2                  0.0               0.0                      0.0   
3                  0.0               0.0                      0.0   
4                  0.0               0.0                      0.0   
5                  0.0               0.0                      0.0   
6                  0.0               0.0                      0.0   
7                  0.0               2.0                      0.0   
8                  0.0               0.0                      0.0   
9                  0.0               0.0                      0.0   

   Actor1KnownGroupCode_count  Actor1Type1Code_count  Actor2Code_count  \
0                         0.0                    0.0               0.0   
1                         0.0                    0.0               0.0   
2                         0.0                    0.0               0.0   
3                         0.0                    0.0               0.0   
4                         0.0                    0.0               0.0   
5                         0.0                    0.0               0.0   
6                         0.0                    0.0               0.0   
7                         0.0                    2.0               0.0   
8                         0.0                    0.0               0.0   
9                         0.0                    0.0               0.0   

   Actor2CountryCode_count  Actor2KnownGroupCode_count  Actor2Type1Code_count  \
0                      0.0                         0.0                    0.0   
1                      0.0                         0.0                    0.0   
2                      0.0                         0.0                    0.0   
3                      0.0                         0.0                    0.0   
4                      0.0                         0.0                    0.0   
5                      0.0                         0.0                    0.0   
6                      0.0                         0.0                    0.0   
7                      0.0                         0.0                    0.0   
8                      0.0         

In [5]:
# Dropping columns before 1989

ucdp.drop(ucdp.loc[ucdp['year']<1989].index, inplace=True)
gdelt.drop(gdelt.loc[gdelt['year']<1989].index, inplace=True)

In [6]:
# Verifying Russia in UCDP

ucdp[['MonthYear','adm_1','country']].loc[ucdp['country']== "Russia (Soviet Union)"]

MonthYear                         adm_1                country
8         198901                Azerbaijan SSR  Russia (Soviet Union)
72        198901                     Uzbek SSR  Russia (Soviet Union)
81        199001                  Armenian SSR  Russia (Soviet Union)
86        199001                Azerbaijan SSR  Russia (Soviet Union)
195       199101                Azerbaijan SSR  Russia (Soviet Union)
...          ...                           ...                    ...
52393     202012  Karachay-Cherkessia republic  Russia (Soviet Union)
52686     202212               Belgorod oblast  Russia (Soviet Union)
52690     202212                Bryansk oblast  Russia (Soviet Union)
52819     202212                 Ryazan oblast  Russia (Soviet Union)
52824     202212                Saratov oblast  Russia (Soviet Union)

[820 rows x 3 columns]

In [7]:
# Verifying Russia in GDELT

gdelt[['year','ActionGeo_ADM1Code','isocode']].loc[gdelt['isocode']== "RUS"]

year ActionGeo_ADM1Code isocode
848983  1989                 RS     RUS
848984  1989                 RS     RUS
848985  1989                 RS     RUS
848986  1989                 RS     RUS
848987  1989                 RS     RUS
...      ...                ...     ...
888532  2023               RSJA     RUS
888533  2023               RSJA     RUS
888534  2023               RSJA     RUS
888535  2023               RSJA     RUS
888536  2023               RSJA     RUS

[33990 rows x 3 columns]

In [8]:
# Verifying Armenia in UCDP

ucdp[['MonthYear','adm_1','country']].loc[ucdp['country']== "Armenia"]

MonthYear                   adm_1  country
3139      201501         Shirak province  Armenia
3151      201501         Tavush province  Armenia
4789      199202         Gorisskiy rayon  Armenia
6906      201202         Tavush province  Armenia
7318      201502         Tavush province  Armenia
9383      199503   Noyemberyanskiy rayon  Armenia
10474     200503         Tavush province  Armenia
11397     201203         Tavush province  Armenia
13234     202303         Ararat province  Armenia
16112     201204         Tavush province  Armenia
19936     200605         Tavush province  Armenia
22278     202105  Geghark’unik’ province  Armenia
22626     202205         Syunik province  Armenia
23136     199406    Krasnosel'skiy rayon  Armenia
25297     201206         Tavush province  Armenia
26991     202206  Geghark’unik’ province  Armenia
27503     199307             Kapan rayon  Armenia
27534     199307   Noyemberyanskiy rayon  Armenia
31273     202107         Ararat province  Armenia
31326     202107  Geghark’unik’ province  Armenia
32038     199308             Kapan rayon  Armenia
32044     199308    Krasnosel'skiy rayon  Armenia
34374     201308         Syunik province  Armenia
35649     202108         Ararat province  Armenia
35701     202108  Geghark’unik’ province  Armenia
39039     201509         Tavush province  Armenia
39523     201809         Tavush province  Armenia
39779     202009  Geghark’unik’ province  Armenia
39935     202109         Ararat province  Armenia
44184     201910         Tavush province  Armenia
44259     202010  Geghark’unik’ province  Armenia
44363     202010         Syunik province  Armenia
48366     202011         Ararat province  Armenia
48504     202011         Syunik province  Armenia
48584     202111  Geghark’unik’ province  Armenia
49180     199212         Gorisskiy rayon  Armenia
49183     199212      Idzhevanskiy rayon  Armenia
49188     199212             Kapan rayon  Armenia
49193     199212    Krasnosel'skiy rayon  Armenia
49218     199212    Shamshadinskiy rayon  Armenia
51792     201612         Tavush province  Armenia
52546     202112  Geghark’unik’ province  Armenia

In [9]:
# Verifying Armenia in GDELT

gdelt[['year','ActionGeo_ADM1Code','isocode']].loc[gdelt['isocode']== "ARM"]

year ActionGeo_ADM1Code isocode
52634  1989                 AM     ARM
52635  1989                 AM     ARM
52636  1989                 AM     ARM
52637  1989                 AM     ARM
52638  1989                 AM     ARM
...     ...                ...     ...
57046  2023               AM11     ARM
57047  2023               AM11     ARM
57048  2023               AM11     ARM
57049  2023               AM11     ARM
57050  2023               AM11     ARM

[4226 rows x 3 columns]

## Exploring UCDP's Admin 1 data

In [10]:
# standardizing the names

ucdp['adm_new'] = ucdp['adm_1'].str.rsplit(n=1).str[0]
ucdp['adm_new'] = ucdp['adm_new'].apply(lambda x: unidecode(x).translate(str.maketrans("", "", string.punctuation)).strip().lower())

ucdp.head(10)

month  year                    adm_1                country  country_id  \
0      1  1989  Agusan del Sur province            Philippines         840   
1      1  1989  Alta Verapaz department              Guatemala          90   
2      1  1989        Ancash department                   Peru         135   
3      1  1989     Antioquia department               Colombia         100   
4      1  1989      Apurímac department                   Peru         135   
5      1  1989      Arequipa department                   Peru         135   
6      1  1989              Assam state                  India         750   
7      1  1989      Ayacucho department                   Peru         135   
8      1  1989           Azerbaijan SSR  Russia (Soviet Union)         365   
9      1  1989         Baghlan province            Afghanistan         700   

     region  best  best_state  best_nonstate  best_onesided  deaths_civilians  \
0      Asia     5         0.0            0.0            5.0                 5   
1  Americas     4         4.0            0.0            0.0                 0   
2  Americas     4         4.0            0.0            0.0                 0   
3  Americas     0         0.0            0.0            0.0                 0   
4  Americas    17        17.0            0.0            0.0                 0   
5  Americas     2         2.0            0.0            0.0                 0   
6      Asia    35        35.0            0.0            0.0                 0   
7  Americas    25        22.0            0.0            3.0                 3   
8    Europe    42         0.0           42.0            0.0                 0   
9      Asia   670       670.0            0.0            0.0                 0   

   number_of_sources  count_conflict_new_id  freq_conflict_new_id  \
0                0.0                      1                   495   
1                0.0                      1                   233   
2                0.0                      1                   292   
3                0.0                      1                   605   
4                0.0                      1                   292   
5                1.0                      1                   292   
6                2.0                      1                   421   
7                0.0                      2                   292   
8                0.0                      1                  4809   
9                0.0                      1                   333   

   count_dyad_new_id  freq_dyad_new_id  MonthYear         adm_new  
0                  1               962     198901  agusan del sur  
1                  1               469     198901    alta verapaz  
2                  1               631     198901          ancash  
3                  1              1072     198901       antioquia  
4                  1               631     198901        apurimac  
5                  1               631     198901        arequipa  
6                  1               884     198901           assam  
7                  2               631     198901        ayacucho  
8                  1              5419     198901      azerbaijan  
9                  1               724     198901         baghlan

In [11]:
# retrieving subdivision isocode from pycountry
def get_admin_iso(df):
    subdivisions = pycountry.subdivisions

    for subdivision in subdivisions:
        subdivision_name = str(subdivision.name)
        subdivision_name = unidecode(subdivision_name).translate(str.maketrans("", "", string.punctuation)).strip().lower()
        
        matching_rows = df[df['adm_new'] == subdivision_name]
        if not matching_rows.empty:
            df.loc[matching_rows.index, 'iso_code'] = subdivision.code

    return df

get_admin_iso(ucdp)
ucdp.head(10)

In [15]:
# exploring unmatched iso codes

print(ucdp['iso_code'].isna().sum()/len(ucdp['iso_code']))
nan_adm_new = ucdp.loc[ucdp['iso_code'].isna(), 'adm_new'].unique().tolist()
len(nan_adm_new)

0.22167776600575148


476

In [33]:
# examples of cases where fuzzy matching must be done

In [34]:
print(pycountry.subdivisions.get(code="PE-APU"))

ucdp[['country','adm_1', 'adm_new','iso_code']].loc[ucdp['adm_new']=="apurimac"]

Subdivision(code='PE-APU', country_code='PE', name='Apurimaq', parent_code=None, type='Region')


country                adm_1   adm_new iso_code
4        Peru  Apurímac department  apurimac      NaN
79       Peru  Apurímac department  apurimac      NaN
8992     Peru  Apurímac department  apurimac      NaN
15706    Peru      Apurímac region  apurimac      NaN
17953    Peru  Apurímac department  apurimac      NaN
20097    Peru      Apurímac region  apurimac      NaN
22649    Peru  Apurímac department  apurimac      NaN
27290    Peru  Apurímac department  apurimac      NaN
40643    Peru  Apurímac department  apurimac      NaN
46558    Peru      Apurímac region  apurimac      NaN

In [35]:
print(pycountry.subdivisions.get(code="ES-MD"))

ucdp[['country','adm_1', 'adm_new','iso_code']].loc[ucdp['adm_new']=="community of"]

Subdivision(code='ES-MD', country_code='ES', name='Madrid, Comunidad de', parent_code=None, type='Autonomous community')


country                adm_1       adm_new iso_code
862     Spain  Community of Madrid  community of      NaN
1166    Spain  Community of Madrid  community of      NaN
4779    Spain  Community of Madrid  community of      NaN
5145    Spain  Community of Madrid  community of      NaN
5220    Spain  Community of Madrid  community of      NaN
9012    Spain  Community of Madrid  community of      NaN
10271   Spain  Community of Madrid  community of      NaN
17965   Spain  Community of Madrid  community of      NaN
18485   Spain  Community of Madrid  community of      NaN
22824   Spain  Community of Madrid  community of      NaN
23004   Spain  Community of Madrid  community of      NaN
23105   Spain  Community of Madrid  community of      NaN
23225   Spain  Community of Madrid  community of      NaN
23836   Spain  Community of Madrid  community of      NaN
27140   Spain  Community of Madrid  community of      NaN
27305   Spain  Community of Madrid  community of      NaN
27593   Spain  Community of Madrid  community of      NaN
28372   Spain  Community of Madrid  community of      NaN
36061   Spain  Community of Madrid  community of      NaN
40561   Spain  Community of Madrid  community of      NaN
40761   Spain  Community of Madrid  community of      NaN
41483   Spain  Community of Madrid  community of      NaN
44803   Spain  Community of Madrid  community of      NaN
45061   Spain  Community of Madrid  community of      NaN
49439   Spain  Community of Madrid  community of      NaN
50058   Spain  Community of Madrid  community of      NaN

In [37]:
print(pycountry.subdivisions.get(code="ES-VC"))

ucdp[['country','adm_1', 'adm_new','iso_code']].loc[ucdp['adm_new']=="valenciana autonomous"]

Subdivision(code='ES-VC', country_code='ES', name='Valenciana, Comunidad', parent_code=None, type='Autonomous community')


country                            adm_1                adm_new iso_code
376     Spain  Valenciana autonomous community  valenciana autonomous      NaN
33012   Spain  Valenciana autonomous community  valenciana autonomous      NaN
36291   Spain  Valenciana autonomous community  valenciana autonomous      NaN
41051   Spain  Valenciana autonomous community  valenciana autonomous      NaN

In [42]:
print(pycountry.subdivisions.get(code="TG-M"))

ucdp[['country','adm_1', 'adm_new','iso_code']].loc[ucdp['adm_new']=="maritime"]

Subdivision(code='TG-M', country_code='TG', name='Maritime (Région)', parent_code=None, type='Region')


country            adm_1   adm_new iso_code
449      Togo  Maritime region  maritime      NaN
969      Togo  Maritime region  maritime      NaN
6088     Togo  Maritime region  maritime      NaN
13696    Togo  Maritime region  maritime      NaN
15153    Togo  Maritime region  maritime      NaN
19773    Togo  Maritime region  maritime      NaN
23536    Togo  Maritime region  maritime      NaN
27419    Togo  Maritime region  maritime      NaN
32533    Togo  Maritime region  maritime      NaN

In [39]:
# examples of cases where matching must be at the country level

ucdp[['country','adm_1', 'adm_new','iso_code']].loc[ucdp['adm_new']=="armenian"]

Subdivision(code='AM-AG', country_code='AM', name='Aragac̣otn', parent_code=None, type='Region')


country         adm_1   adm_new iso_code
81     Russia (Soviet Union)  Armenian SSR  armenian      NaN
8800   Russia (Soviet Union)  Armenian SSR  armenian      NaN
18151  Russia (Soviet Union)  Armenian SSR  armenian      NaN
31740  Russia (Soviet Union)  Armenian SSR  armenian      NaN

## Exploring GDELT's Admin 1 data

In [43]:
# retrieving subdivision names from FIPS codes

print("\n****************\n\nGetting names from FIPS code...\n")
fips = pd.read_csv(path +'/data/fipsadm.csv') # codes' list

fips['adm_new'] = fips['region_name'].apply(lambda x: unidecode(x).translate(str.maketrans("", "", string.punctuation)).strip().lower())

fips.head(10)


****************

Getting names from FIPS code...



region_code region_division          region_name              adm_new
0        AA00         country                ARUBA                aruba
1        AC00         country  ANTIGUA AND BARBUDA  antigua and barbuda
2        AC01      dependency              Barbuda              barbuda
3        AC03          parish         Saint George         saint george
4        AC04          parish           Saint John           saint john
5        AC05          parish           Saint Mary           saint mary
6        AC06          parish           Saint Paul           saint paul
7        AC07          parish          Saint Peter          saint peter
8        AC08          parish         Saint Philip         saint philip
9        AC09      dependency              Redonda              redonda

In [44]:
# merging them:

gdelt['ActionGeo_ADM1Code'] = gdelt['ActionGeo_ADM1Code'].apply(lambda x: x + '00' if len(x) == 2 else x)
gdelt = pd.merge(gdelt,fips,how='left', left_on = 'ActionGeo_ADM1Code', right_on = 'region_code')

# removing useless variables:
gdelt = gdelt.drop(columns=['region_code', 'region_name'])

gdelt.head(10)

ActionGeo_ADM1Code isocode  year  month  count_events_1  count_events_2  \
0               AA00     ABW  1989      3             0.0             0.0   
1               AA00     ABW  1989      8             2.0             0.0   
2               AA00     ABW  1990      6             0.0             0.0   
3               AA00     ABW  1991      4             0.0             0.0   
4               AA00     ABW  1992     12             0.0             0.0   
5               AA00     ABW  1993      6             0.0             2.0   
6               AA00     ABW  1993      7             2.0             0.0   
7               AA00     ABW  1993     10             0.0             0.0   
8               AA00     ABW  1993     12             0.0             4.0   
9               AA00     ABW  1994      7             0.0             0.0   

   count_events_3  count_events_4  count_events_5  count_events_6  ...  \
0             4.0             0.0             0.0             0.0  ...   
1             0.0             0.0             0.0             0.0  ...   
2             0.0             8.0             0.0             0.0  ...   
3             0.0             0.0             4.0             0.0  ...   
4             0.0             4.0             0.0             0.0  ...   
5             0.0             8.0             0.0             0.0  ...   
6             0.0             0.0             0.0             0.0  ...   
7             2.0             0.0             0.0             0.0  ...   
8             0.0             0.0             0.0             0.0  ...   
9             0.0             8.0             0.0             0.0  ...   

   Actor1CountryCode_count  Actor1KnownGroupCode_count  Actor1Type1Code_count  \
0                      0.0                         0.0                    0.0   
1                      0.0                         0.0                    0.0   
2                      0.0                         0.0                    0.0   
3                      0.0                         0.0                    0.0   
4                      0.0                         0.0                    0.0   
5                      0.0                         0.0                    0.0   
6                      0.0                         0.0                    0.0   
7                      0.0                         0.0                    0.0   
8                      0.0                         0.0                    0.0   
9                      0.0                         0.0                    0.0   

   Actor2Code_count  Actor2CountryCode_count  Actor2KnownGroupCode_count  \
0               0.0                      0.0                         0.0   
1               0.0                      0.0                         0.0   
2               0.0                      0.0                         0.0   
3               0.0                      0.0                         0.0   
4               0.0                      0.0                         0.0   
5               0.0                      0.0                         0.0   
6               0.0                      0.0                         0.0   
7               0.0                      0.0                         0.0   
8               0.0                      0.0                         0.0   
9               0.0                      0.0                         0.0   

   Actor2Type1Code_count  AvgGoldsteinScale  region_division  adm_new  
0                    0.0                NaN          country    aruba  
1                    0.0                NaN          country    aruba  
2                    0.0                NaN          country    aruba  
3                    0.0                NaN          country    aruba  
4                    0.0                NaN          country    aruba  
5                    0.0                NaN          country    aruba  
6                    0.0                NaN          country    aruba  
7                    0.0          

In [45]:
# exploring unmatched names

print(gdelt['adm_new'].isna().sum()/len(gdelt['adm_new']))

nans_gdelt = gdelt.loc[gdelt['adm_new'].isna(), 'ActionGeo_ADM1Code'].unique().tolist()
len(nans_gdelt)

0.11023792158077558


699

In [46]:
# examples of cases where code does not match

In [57]:
print(fips[['region_code','region_name','adm_new']].loc[fips['region_code']=='AO10'])
print(fips.loc[fips['adm_new']=='luanda'])

## AO10 no longer exists, but used to refer to Luanda, which is now AO20

print(gdelt[['ActionGeo_ADM1Code','isocode','adm_new']].loc[gdelt['ActionGeo_ADM1Code']=='AO10'])
print(gdelt[['ActionGeo_ADM1Code','isocode','adm_new']].loc[gdelt['ActionGeo_ADM1Code']=='AO20'])

Empty DataFrame
Columns: [region_code, region_name, adm_new]
Index: []
    region_code     region_division region_name adm_new
225        AO20  province provincia      Luanda  luanda
      ActionGeo_ADM1Code isocode adm_new
56287               AO10     AGO     NaN
56288               AO10     AGO     NaN
56289               AO10     AGO     NaN
56290               AO10     AGO     NaN
56291               AO10     AGO     NaN
...                  ...     ...     ...
56695               AO10     AGO     NaN
56696               AO10     AGO     NaN
56697               AO10     AGO     NaN
56698               AO10     AGO     NaN
56699               AO10     AGO     NaN

[413 rows x 3 columns]
      ActionGeo_ADM1Code isocode adm_new
59674               AO20     AGO  luanda
59675               AO20     AGO  luanda
59676               AO20     AGO  luanda
59677               AO20     AGO  luanda
59678               AO20     AGO  luanda
...                  ...     ...     ...
59990         

In [63]:
print(fips[['region_code','region_name','adm_new']].loc[fips['region_code']=='BE13'])
print(fips[['region_code','region_name','adm_new']].loc[fips['region_code']=='BE08'])
print(fips[['region_code','region_name','adm_new']].loc[fips['region_code']=='BE09'])
print(fips[['region_code','region_name','adm_new']].loc[fips['region_code']=='BE12'])

## BE13 no longer exists but used to refer to the Flanders region, which is not subdivided in GDELT but is in FIPS

print(gdelt[['ActionGeo_ADM1Code','isocode','adm_new']].loc[gdelt['ActionGeo_ADM1Code']=='BE13'])
print(gdelt[['ActionGeo_ADM1Code','isocode','adm_new']].loc[gdelt['ActionGeo_ADM1Code']=='BE08'])
print(gdelt[['ActionGeo_ADM1Code','isocode','adm_new']].loc[gdelt['ActionGeo_ADM1Code']=='BE09'])
print(gdelt[['ActionGeo_ADM1Code','isocode','adm_new']].loc[gdelt['ActionGeo_ADM1Code']=='BE12'])

Empty DataFrame
Columns: [region_code, region_name, adm_new]
Index: []
    region_code      region_name         adm_new
333        BE08  Oost-Vlaanderen  oostvlaanderen
    region_code      region_name         adm_new
334        BE09  West-Vlaanderen  westvlaanderen
    region_code     region_name        adm_new
337        BE12  Vlamms-Brabant  vlammsbrabant
      ActionGeo_ADM1Code isocode adm_new
87614               BE13     BEL     NaN
87615               BE13     BEL     NaN
87616               BE13     BEL     NaN
87617               BE13     BEL     NaN
87618               BE13     BEL     NaN
...                  ...     ...     ...
88011               BE13     BEL     NaN
88012               BE13     BEL     NaN
88013               BE13     BEL     NaN
88014               BE13     BEL     NaN
88015               BE13     BEL     NaN

[402 rows x 3 columns]
Empty DataFrame
Columns: [ActionGeo_ADM1Code, isocode, adm_new]
Index: []
Empty DataFrame
Columns: [ActionGeo_ADM1Code, iso

In [65]:
print(fips[['region_code','region_name','adm_new']].loc[fips['region_code']=='BE14'])
print(fips[['region_code','region_name','adm_new']].loc[fips['region_code']=='BE03'])
print(fips[['region_code','region_name','adm_new']].loc[fips['region_code']=='BE04'])
print(fips[['region_code','region_name','adm_new']].loc[fips['region_code']=='BE06'])
print(fips[['region_code','region_name','adm_new']].loc[fips['region_code']=='BE07'])
print(fips[['region_code','region_name','adm_new']].loc[fips['region_code']=='BE10'])

## BE14 no longer exists but used to refer to Wallonia (now subdivided in FIPS but not in GDELT)

print(gdelt[['ActionGeo_ADM1Code','isocode','adm_new']].loc[gdelt['ActionGeo_ADM1Code']=='BE14'])
print(gdelt[['ActionGeo_ADM1Code','isocode','adm_new']].loc[gdelt['ActionGeo_ADM1Code']=='BE03'])
print(gdelt[['ActionGeo_ADM1Code','isocode','adm_new']].loc[gdelt['ActionGeo_ADM1Code']=='BE04'])
print(gdelt[['ActionGeo_ADM1Code','isocode','adm_new']].loc[gdelt['ActionGeo_ADM1Code']=='BE06'])
print(gdelt[['ActionGeo_ADM1Code','isocode','adm_new']].loc[gdelt['ActionGeo_ADM1Code']=='BE07'])
print(gdelt[['ActionGeo_ADM1Code','isocode','adm_new']].loc[gdelt['ActionGeo_ADM1Code']=='BE10'])

Empty DataFrame
Columns: [region_code, region_name, adm_new]
Index: []
    region_code region_name  adm_new
328        BE03     Hainaut  hainaut
    region_code region_name adm_new
329        BE04       Liege   liege
    region_code region_name     adm_new
331        BE06  Luxembourg  luxembourg
    region_code region_name adm_new
332        BE07       Namur   namur
    region_code     region_name         adm_new
335        BE10  Brabant Wallon  brabant wallon
      ActionGeo_ADM1Code isocode adm_new
88016               BE14     BEL     NaN
88017               BE14     BEL     NaN
88018               BE14     BEL     NaN
88019               BE14     BEL     NaN
88020               BE14     BEL     NaN
...                  ...     ...     ...
88403               BE14     BEL     NaN
88404               BE14     BEL     NaN
88405               BE14     BEL     NaN
88406               BE14     BEL     NaN
88407               BE14     BEL     NaN

[392 rows x 3 columns]
Empty DataFrame
Col

In [66]:
# in ucdp, the division is the same as in gdelt though, which is good

ucdp[['country','adm_1', 'adm_new','iso_code']].loc[ucdp['country']=="Belgium"]

country                    adm_1          adm_new iso_code
11904  Belgium  Brussels-Capital region  brusselscapital      NaN
21789  Belgium          Wallonia region         wallonia      NaN
34887  Belgium          Wallonia region         wallonia      NaN

In [ ]:
# examples of cases where code is missing

In [80]:
gdelt[['ActionGeo_ADM1Code','isocode','adm_new']].loc[gdelt['ActionGeo_ADM1Code']=='LGA1']

print(fips[['region_code','region_name','adm_new']].loc[fips['adm_new']=='latvia'])

# nothing for subdivisions of Latvia whatsoever

     region_code region_name adm_new
1789        LG00      LATVIA  latvia


In [78]:
ucdp[['country','adm_1', 'adm_new','iso_code']].loc[ucdp['country']=="Latvia"]

# no UCDP entries for Latvia either

Empty DataFrame
Columns: [country, adm_1, adm_new, iso_code]
Index: []

## Merging UCDP and GDELT at the Admin 1 level

In [97]:
merged_df = gdelt.merge(ucdp, on='adm_new', how='outer', indicator=True)

In [99]:
len(merged_df.loc[merged_df['_merge'] == 'left_only', 'adm_new'].unique())

2585

In [100]:
len(merged_df.loc[merged_df['_merge'] == 'right_only', 'adm_new'].unique())

637

In [104]:
merged_df.loc[merged_df['_merge'] == 'right_only', 'adm_new'].unique()

array(['battambang', 'beirut', 'bougainville', 'kabul', 'kandahar',
       'kunduz', 'metropolitan manila', 'natal', 'north central',
       'northern darfur', 'paktya', 'sabaragamuwa', 'south lebanon',
       'tigray', 'transvaal', 'upper nile', 'uzbek', 'west',
       'zamboanga del sur', 'aceh', 'armenian', 'assiut', 'begemder',
       'cape', 'distrito', 'kalingaapayao', 'kayin', 'kordofan',
       'north eastern', 'orange free state', 'pais vasco autonomous',
       'shan', 'siem reap oddar meanchey', 'timor timur', 'uva',
       'vientiane', 'welega', 'western equatoria', 'wollo', 'xaisomboun',
       'xiangkhoang', 'attamim', 'azad jammu', 'banteay meanchey',
       'bogota', 'byumba', 'galgudud', 'georgian', 'gisenyi', 'kachin',
       'kaffa', 'latvian', 'lithuanian', 'mbale', 'riyadh', 'ruhengeri',
       'bahr al ghazal', 'brodposavina', 'cataluna autonomous',
       'dakhlet nouadhibou', 'darfur', 'delhi union', 'equatoria',
       'far north', 'khartoum', 'khojaly', 'kinsh

In [108]:
# regions where the names have different spelling

In [109]:
ucdp[['country','adm_1', 'adm_new','iso_code']].loc[ucdp['adm_new']=="beirut"]

country            adm_1 adm_new iso_code
11     Lebanon  Beirut province  beirut      NaN
89     Lebanon  Beirut province  beirut      NaN
2862   Lebanon  Beirut province  beirut      NaN
4534   Lebanon  Beirut province  beirut      NaN
4612   Lebanon  Beirut province  beirut      NaN
8721   Lebanon  Beirut province  beirut      NaN
8809   Lebanon  Beirut province  beirut      NaN
13452  Lebanon  Beirut province  beirut      NaN
13546  Lebanon  Beirut province  beirut      NaN
13928  Lebanon  Beirut province  beirut      NaN
14152  Lebanon  Beirut province  beirut      NaN
17957  Lebanon  Beirut province  beirut      NaN
18052  Lebanon  Beirut province  beirut      NaN
19365  Lebanon  Beirut province  beirut      NaN
22657  Lebanon  Beirut province  beirut      NaN
22733  Lebanon  Beirut province  beirut      NaN
25479  Lebanon  Beirut province  beirut      NaN
27134  Lebanon  Beirut province  beirut      NaN
28940  Lebanon  Beirut province  beirut      NaN
31665  Lebanon  Beirut province  beirut      NaN
31748  Lebanon  Beirut province  beirut      NaN
33406  Lebanon  Beirut province  beirut      NaN
34272  Lebanon  Beirut province  beirut      NaN
36045  Lebanon  Beirut province  beirut      NaN
40443  Lebanon  Beirut province  beirut      NaN
47564  Lebanon  Beirut province  beirut      NaN
48926  Lebanon  Beirut province  beirut      NaN
49337  Lebanon  Beirut province  beirut      NaN

In [110]:
gdelt[['ActionGeo_ADM1Code','isocode','adm_new']].loc[gdelt['adm_new']=='beyrouth']

ActionGeo_ADM1Code isocode   adm_new
498857               LE04     LBN  beyrouth
498858               LE04     LBN  beyrouth
498859               LE04     LBN  beyrouth
498860               LE04     LBN  beyrouth
498861               LE04     LBN  beyrouth
...                   ...     ...       ...
499265               LE04     LBN  beyrouth
499266               LE04     LBN  beyrouth
499267               LE04     LBN  beyrouth
499268               LE04     LBN  beyrouth
499269               LE04     LBN  beyrouth

[413 rows x 3 columns]

##  Issues to deal with:

* UCDP:
    * Soviet Union is a country until 1991 and each Republic is an Admin 1 region. After 1991, each Republic is its own country with smaller Admin 1 regions.
    * 22% of the data is missing an ISO code (approx. 485 ISO codes) from pycountry.
    
* GDELT:
    * 11% of the data is missing a name tied to the FIPS code (approx. 699 names) from the FIPS.csv file

* Merging:
    * How do we merge country-level GDELT information with admin-level deaths?